apply.c

```diff
@@ -1591,15 +1595,24 @@ static int find_header(struct apply_state *state,
 		 * or mode change, so we handle that specially
 		 */
 		if (!memcmp("diff --git ", line, 11)) {
-			int git_hdr_len = parse_git_diff_header(&state->root, &state->linenr,
+			int git_hdr_len;
+			struct promise_t *parse_git_diff_header_promise = promise_init();
+			parse_git_diff_header(&state->root, &state->linenr,
 								state->p_value, line, len,
-								size, patch);
-			if (git_hdr_len < 0)
-				return -128;
+								size, patch, parse_git_diff_header_promise);
+			promise_assert_finished(parse_git_diff_header_promise);
+
+			if (parse_git_diff_header_promise->state == PROMISE_FAILURE) {
+				PROMISE_BUBBLE_UP(promise, parse_git_diff_header_promise,
+						 _("could not find file diff header"));
+			}
+
+			git_hdr_len = promise->result.success_result;
```

Comment:
```suggestion
			git_hdr_len = parse_git_diff_header_promise->result.success_result;
```

Without this, you see tons of test failures.

This kind of finicky issue (where it is too easy to have perfectly compiling code even though the _wrong_ promise object was used by mistake) points to a problem with this approach that you may want to mention in the cover letter: this approach requires a lot of repetitions.

Reply:
Great point. Do you have any ideas by chance for how to reduce the repetition?

Reply:
You could use the comma operator to combine the conditional return with the assignment of the return value but elegant it ain't...